multiqc/utils/search_patterns.yaml

```diff
@@ -17,6 +17,14 @@ bamtools/stats:
   contents: "Stats for BAM file(s):"
   shared: true
   num_lines: 10
+bases2fastq/persample:
+  fn: "*_stats.json"
+bases2fastq/run:
+  fn: "RunStats.json"
+bases2fastq/project:
+  fn: "*_RunStats.json"
```

Comment:
These filename patterns are not very specific, especially `*_stats.json`. Is there anything else that we can also filter on? More specific is good, as it'll return a smaller list of files to the module (faster) and also because when a file is found it's "consumed", meaning if it's for another tool then it won't be passed to that module.

Can we guarantee that these files always start with the string `{"AnalysisVersion":`? Could then combine both `fn` and `content` to make a more specific (`AND`) filter.

See Bcl2fast for a similar example:
https://github.com/ewels/MultiQC/blob/41c1c85813c27a6cec2158d664594afacf178417/multiqc/utils/search_patterns.yaml#L89-L92

Reply:
Our jsons can be quite large - we try to give the customer an abundance of metrics at the run, sample, and cycle level.
Recently we've moved to flatten our json (e.g. jq -c) to reduce file size, which then causes the jsons to only have a single (flattened) line :)

For our sample specific jsons, the `fn: "*_stats.json"` will only ever match to only file.  The * is there to capture the <sample_id> that we use to prefix these file names.  It may be better to not search for a contents across N lines given the above?  Thoughts?

Reply:
> Will this cause issues on the multiqc side given that now you must slurp in the entire json to then search for our `content` pattern?

It'll slow it down for sure, though the `fn` pattern is there at least so it hopefully wouldn't be so many files.

This means that the only remaining reason for doing this really is in case this search pattern is preventing files from being found by other tools. I think it's still worth adding it in, if that's ok.

@vladsavelyev another search optimisation that we could do is to read in chunks of bytes to check for `contents` rather than just counting lines, to avoid this situation of having to read a 100MB file with a single line, just to match the first 10 characters. Pretty rare edge case though, so maybe not worth the effort.

Reply:
ok - marking this as resolved.  We can look into not completely flattening the entire json to help improve this.